nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -2718,4 +2743,219 @@ pub mod test {
+            let orig_nexus_filesystem_pool =
+                nexus.filesystem_pool.clone().expect(
+                    "nexus instances in new blueprints \
+                     should have a filesystem_pool",
+                );
+
+            match sled_ids.len() {
+                0 => {
+                    expected_zones_config_gen.push(zones_config.generation);
+                }
+                1 | 2 => {
+                    expected_zones_config_gen
+                        .push(zones_config.generation.next());
+                    nexus.filesystem_pool = None;
+                }
+                _ => unreachable!("unexpected number of sleds in test"),
+            }
+            sled_ids.push(*sled_id);
+            expected_filesystem_pools.push(orig_nexus_filesystem_pool.clone());
+        }
+
+        // Create a builder and produce a new blueprint.
+        let blueprint1 = {
+            let mut builder = BlueprintBuilder::new_based_on(
+                &logctx.log,
+                &parent,
+                &input,
+                &collection,
+                "test",
+            )
+            .expect("constructed builder");
+
+            for sled_id in input.all_sled_ids(SledFilter::InService) {
+                builder
+                    .sled_ensure_zone_datasets(sled_id)
+                    .expect("ensured zone datasets");
+            }
+
+            builder.build()
+        };
+
+        // Check that our backfilling was correct.
+        for (i, &sled_id) in sled_ids.iter().enumerate() {
+            let zones_config =
+                blueprint1.blueprint_zones.get(&sled_id).expect("found sled");
+            assert_eq!(
+                zones_config.generation, expected_zones_config_gen[i],
+                "unexpected generation on sled {i}"
+            );
+
+            let nexus = zones_config
+                .zones
+                .iter()
+                .find(|z| z.zone_type.is_nexus())
+                .expect("found nexus on sled");
+            assert_eq!(
+                nexus.filesystem_pool.as_ref(),
+                Some(&expected_filesystem_pools[i]),
+                "unexpected filesystem_pool on sled {i}"
+            );
+        }
+
+        // Check that the new blueprint is blippy-clean.
+        verify_blueprint(&blueprint1);
+
+        // It's possible a sled-agent could restart in between when the
+        // inventory collection we used for backfilling was created and when we
+        // try to send it the new zone configs, and in doing so it could have
+        // changed the zpool it chose for any zone type that doesn't have a
+        // durable dataset (e.g., Nexus). To emulate this, mutate our collection
+        // and change the zpool for sled 2's Nexus to a different zpool; our
+        // backfilling should correct this again, even though the parent
+        // blueprint has a non-`None` filesystem pool for this Nexus.
```

Comment:
Nice job with this test; I know this was a bad "hypothetical", it's great to see it being covered so explicitly.